File: python-mcp-workspace/mongo_mcp_server/handlers/database_handlers.py
from ..config import DEFAULT_DATABASE_NAME
from ..mongo import get_database, list_database_names
from ..serializers import to_jsonable
from ..validators import (
    validate_collection_name,
    validate_confirmed,
    validate_create_collection_options,
    validate_database_name,
)


def list_databases(include_collections: bool = True) -> dict:
    databases = list_database_names()

    if not include_collections:
        return {
            "databases": databases,
        }

    return {
        "databases": [
            {
                "database": database,
                "collections": get_database(database).list_collection_names(),
            }
            for database in databases
        ],
    }


def list_collections(database: str = DEFAULT_DATABASE_NAME) -> dict:
    collections = get_database(database).list_collection_names()

    return {
        "database": database,
        "collections": collections,
    }


def create_database(database: str, collection: str = "_mcp_init") -> dict:
    validate_database_name(database)
    validate_collection_name(collection)

    database_instance = get_database(database)
    existing_collections = database_instance.list_collection_names(
        filter={"name": collection}
    )

    if not existing_collections:
        database_instance.create_collection(collection)

    return {
        "database": database,
        "collection": collection,
        "created": not bool(existing_collections),
    }


def create_collection(
    collection: str,
    database: str = DEFAULT_DATABASE_NAME,
    options: dict | None = None,
) -> dict:
    options = options or {}

    validate_collection_name(collection)
    validate_create_collection_options(options)

    get_database(database).create_collection(collection, **options)

    return {
        "database": database,
        "collection": collection,
        "created": True,
    }


def drop_database(database: str, confirm: bool) -> dict:
    validate_database_name(database)
    validate_confirmed(confirm)

    get_database(database).client.drop_database(database)

    return {
        "database": database,
        "dropped": True,
    }


def drop_collection(
    collection: str,
    confirm: bool,
    database: str = DEFAULT_DATABASE_NAME,
) -> dict:
    validate_collection_name(collection)
    validate_confirmed(confirm)

    result = get_database(database).drop_collection(collection)

    return to_jsonable(
        {
            "database": database,
            "collection": collection,
            "dropped": True,
            "result": result,
        }
    )

File: mongo-mcp-server/src/handlers/databaseHandlers.js
import { defaultDatabaseName } from "../config.js";
import { getDatabase, validateDatabaseName } from "../mongo.js";
import { textContent } from "../response.js";
import {
  validateCollectionName,
  validateConfirmed,
  validateCreateCollectionOptions
} from "../validators.js";

export const databaseHandlers = {
  list_collections: async ({ database = defaultDatabaseName }) => {
    const cols = await getDatabase(database).listCollections().toArray();

    return textContent({
      database,
      collections: cols.map((collection) => collection.name)
    });
  },

  create_database: async ({ database, collection = "_mcp_init" }) => {
    validateDatabaseName(database);
    validateCollectionName(collection);

    const databaseInstance = getDatabase(database);
    const existingCollections = await databaseInstance
      .listCollections({ name: collection })
      .toArray();

    if (existingCollections.length === 0) {
      await databaseInstance.createCollection(collection);
    }

    return textContent({
      database,
      collection,
      created: existingCollections.length === 0
    });
  },

  create_collection: async ({
    database = defaultDatabaseName,
    collection,
    options = {}
  }) => {
    validateCollectionName(collection);
    validateCreateCollectionOptions(options);

    await getDatabase(database).createCollection(collection, options);

    return textContent({
      database,
      collection,
      created: true
    });
  },

  drop_database: async ({ database, confirm }) => {
    validateDatabaseName(database);
    validateConfirmed(confirm);

    const result = await getDatabase(database).dropDatabase();

    return textContent({
      database,
      dropped: result
    });
  },

  drop_collection: async ({
    database = defaultDatabaseName,
    collection,
    confirm
  }) => {
    validateCollectionName(collection);
    validateConfirmed(confirm);

    const result = await getDatabase(database).dropCollection(collection);

    return textContent({
      database,
      collection,
      dropped: result
    });
  }
};
